File: modules/wem-admin-ui/src/main/resources/web/admin/common/js/content/ContentName.ts
module api.content {

    export class ContentName extends api.Name implements api.Equitable {

        public static UNNAMED_PREFIX: string = "__unnamed__";


        constructor(name: string) {
            super(name);
        }

        isUnnamed(): boolean {
            return false;
        }

        toUnnamed(): ContentUnnamed {
            api.util.assert(api.ObjectHelper.iFrameSafeInstanceOf(this, ContentUnnamed), "this is not a ContentUnnamed");
            return <ContentUnnamed>this;
        }

        equals(o: api.Equitable): boolean {

            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, ContentName)) {
                return false;
            }

            if (!super.equals(o)) {
                return false;
            }

            return true;
        }

        public static fromString(str: string): ContentName {

            api.util.assert(str != null, "name cannot be null");
            if (str.indexOf(ContentName.UNNAMED_PREFIX) == 0) {
                return new ContentUnnamed(str);
            }
            else {
                return new ContentName(str);
            }
        }

    }
}

File: modules/core-api/src/main/java/com/enonic/xp/content/ContentName.java
package com.enonic.xp.content;


import java.util.UUID;

import com.google.common.base.Preconditions;

import com.enonic.xp.name.Name;

public class ContentName
    extends Name
{
    protected ContentName( final String name )
    {
        super( name );
    }

    public boolean isUnnamed()
    {
        return false;
    }

    private final static String UNNAMED_PREFIX = "__unnamed__";

    public static class Unnamed
        extends ContentName
    {
        public Unnamed( final String name )
        {
            super( name );
            Preconditions.checkArgument( name.startsWith( UNNAMED_PREFIX ),
                                         "An UnnamedContent must start with [" + UNNAMED_PREFIX + ": ]" + name );
        }

        @Override
        public boolean isUnnamed()
        {
            return true;
        }

        public static ContentName withUniqueness()
        {
            return new Unnamed( UNNAMED_PREFIX + UUID.randomUUID().toString() );
        }

        public boolean hasUniqueness()
        {
            return toString().length() > UNNAMED_PREFIX.length();
        }
    }

    public static ContentName from( final String name )
    {
        Preconditions.checkNotNull( name, "Cannot resolve ContentName from null" );
        if ( name.startsWith( UNNAMED_PREFIX ) )
        {
            return new Unnamed( name );
        }
        else
        {
            return new ContentName( name );
        }
    }
}
